README.md
# Onyx Server

Mailboxing & data service for the [Onyx](https://github.com/MainframeHQ/onyx) app.

## Prerequisites

[Node](https://nodejs.org/en/) v8+ with [npm](https://www.npmjs.com/).

## Installation

```sh
npm install --global onyx-server
```

## Usage

### CLI

```sh
onyx-server --port 5000 --http-url http://localhost:8500 --ws-url ws://localhost:8546
```

All arguments are optional, when not provided the server will use environment
variables `ONYX_PORT`, `SWARM_HTTP_URL` and `SWARM_HTTP_URL` or its defaults
(WebSocket on `ws://localhost:8546`, HTTP on `http://localhost:8500` and port
5000).

Additionally you can pass `-u` or `--unsecure` to dismiss using TLS, only recommended
for when connecting client and server over a local connection

The `DEBUG` environment variable can be used to activate logs, ex:

```sh
DEBUG="onyx*" onyx-server
```

### Connection security

The Onyx client connects to Onyx server using a WebSocket, and so uses TLS certificates to authenticate and encrypt the connection.
Use of client certificates is enforced so that only your clients with the correct certificate will be allowed to connect to the server, others are rejected.
For mobile clients, where it can be more difficult to handle cert files, the server allows clients accessing the cert endpoint to download a password-encrypted p12 client cert.
For convenience, you can use the provided script to generate a set of unique self-signed certificates.

```sh
./scripts/gen-certs.sh -p <certificate-password> -i <ip-address-to-certify> -d <domain-to-certify>
```

### Setting up an Onyx server on AWS

You can use a pre-built [AMI](https://en.wikipedia.org/wiki/Amazon_Machine_Image)
to conveniently set up an Onyx server on AWS.

To do it, make sure you have an AWS account and your AWS CLI is configured to
use the `eu-west-1` (Ireland) region as default. We're going to assume you have a
[VPC](https://eu-west-1.console.aws.amazon.com/vpc/home?region=eu-west-1#)
configured in that region.

Aside from a VPC and an Internet Gateway, the
[Route Table](https://eu-west-1.console.aws.amazon.com/vpc/home?region=eu-west-1#routetables:)
has to be configured for this VPC, with routes set like these:

| Destination | Target                      | Status | Propagated |
| ---         | ----                        | ---    | ---        |
| 10.0.0.0/16 | local                       | Active | No         |
| 0.0.0.0/0   | \<INTERNET GATEWAY ID HERE> | Active | No         |

#### 1. Create a security group for your Onyx Server

Go to [security group management](https://eu-west-1.console.aws.amazon.com/ec2/v2/home?region=eu-west-1#SecurityGroups:sort=groupId)
in the AWS dashboard and create a new security group. Make sure you're creating
it in the right vpc. Set the following group rules:

**Inbound**

| Type            | Protocol | Port Range | Source    | Description    |
| ---             | ---      |        --- | ---       | ---            |
| SSH (22)        | TCP      |         22 | 0.0.0.0/0 | SSH            |
| Custom TCP Rule | TCP      |      30399 | 0.0.0.0/0 | swarm TCP      |
| Custom UDP Rule | UDP      |      30399 | 0.0.0.0/0 | swarm UDP      |
| Custom TCP Rule | TCP      |       5000 | 0.0.0.0/0 | Onyx interface |
| Custom TCP Rule | TCP      |       5002 | 0.0.0.0/0 | cert endpoint  |

**Outbound**

| Type        | Protocol | Port Range | Source    | Description |
| ---         | ---      | ---        | ---       | ---         |
| ALL Traffic | ALL      | ALL        | 0.0.0.0/0 | ALL Traffic |

#### 2. Create a subnet for your Onyx server

Go to [subnet management](https://eu-west-1.console.aws.amazon.com/vpc/home?region=eu-west-1#subnets:)
in the AWS dashboard and create a new subnet in your VPC. Make sure it's within
the vpc CIDR range. For example if the VPC CIDR is `10.0.0.0/16`, the sg
IPv4 CIDR block can be `10.0.1.0/24`.

#### 3. Create an SSH key
You're going to use it to connect to the Onyx server EC2 node.

```bash
$ mkdir ~/ssh
$ aws ec2 create-key-pair --key-name my_key --output text --query KeyMaterial > ~/ssh/my_key.pem
$ chmod 400 ~/ssh/my_key.pem
```

#### 4. Launch the Onyx server

Make sure you have [AWS CLI](https://aws.amazon.com/cli/) installed and configured.
In terminal, run the following command:

Find your security group and subnet ids, which should be something like `sg-XXXXXXXX`
and `subnet-XXXXXXXX`, respectively. Use them in the following command

```bash
$ aws ec2 run-instances \
    --image-id ami-1a066763 \
    --instance-type t2.micro \
    --key-name my_key \
    --security-group-ids <SG ID HERE> \
    --subnet-id <SUBNET ID HERE> \
    --tag-specification "ResourceType=instance,Tags=[{Key=Name,Value=my_onyx_node}]" \
    --associate-public-ip-address
```

This will launch your personal Onyx server. It will generate an
account for you.

#### 5. Fetch the certificates from the node

In order to connect to the server, the client will need to use the right
certificates - otherwise the connection will be rejected. They are generated on
the server and you need to fetch them first.

Find the public IP of the node you created
[here](https://eu-west-1.console.aws.amazon.com/ec2/v2/home?region=eu-west-1#Instances:sort=instanceId)
and copy the relevant files from it:

```bash
$ scp -i ~/ssh/my_key.pem ubuntu@<NODE PUBLIC IP HERE>:"~/certs/ca-crt.pem ~/certs/client-crt.pem ~/certs/client-key.pem" .
```

#### 6. Connect to your Onyx server

Launch Onyx and as the `Onyx server websocket url` use
`wss://<NODE PUBLIC IP HERE>:5000/graphql`. When prompted for the certificates
use the one you downloaded in the previous step.

You're connected!

### Development

To build local version run `yarn start`. Afterwards you can start the built server
from `./bin/onyx-server`.

Onyx server depends on having a local swarm node running. You can start it by running
the `start_swarm_node.sh` script. This should allow you to run `onyx-server` with
no special arguments.

in one shell:
```sh
./scripts/start_swarm_node.sh <some_swarm_data_directory_here>
```

in another shell:
```sh
yarn start
./bin/onyx-server
```

### API

```js
import Conf from 'conf'
import startServer from 'onyx-server'

startServer({
  httpUrl: 'http://localhost:8500',
  wsUrl: 'ws://localhost:8546',
  port: 5000,
  store: new Conf(),
}).then(
  server => {
    console.log('server started')
  },
  err => {
    console.log('failed to start server', err)
  },
)
```

All parameters are optional, fallback values will be used for the parameters not
provided.

## License

MIT.\
See [LICENSE](LICENSE) file.


CHANGELOG.md
## v0.5.0 (in progress)

Require staking Mainframe tokens to start.

## v0.4.0 (2017-12-11)

Breaking PSS API changes to hex values.

## v0.3.0 (2017-12-11)

Return `server` object when server successfully starts.

## v0.2.0 (2017-12-07)

Added support for TLS certificates.

## v0.1.0 (2017-11-17)

First release.


src/index.js
// @flow

import type Conf from 'conf'

import { pubKeyToAddress } from './crypto'
import DB from './db'
import {
  setupPss,
  setupContactTopic,
  subscribeToStoredConvos,
} from './pss/client'
import createServer from './server'
import createContracts from './contracts'

const { ONYX_PORT, SWARM_HTTP_URL, SWARM_WS_URL, WEB3_URL } = process.env

const ENS_STAKE_NAME = {
  TESTNET: 'stake.mainframe.test', // ROPSTEN
  MAINNET: 'stake.mainframehq.eth',
}

const ENS_ADDRESSES = {
  TESTNET: '0x112234455c3a32fd11230c42e7bccd4a84e02010', // ROPSTEN
  MAINNET: '0x314159265dD8dbb310642f98f50C066173C1259b',
}

const WEB3_URLS = {
  TESTNET: 'https://ropsten.infura.io/KWLG1YOMaYgl4wiFlcJv',
  MAINNET: 'https://mainnet.infura.io/KWLG1YOMaYgl4wiFlcJv',
}

type Options = {
  wsUrl?: string,
  httpUrl?: string,
  store?: Conf,
  port?: number,
  unsecure?: boolean,
  certsDir?: string,
  testNet?: boolean,
  web3Url?: string,
  stakeEnsAddress?: string,
  ensAddress?: string,
}

const start = async (opts: Options) => {
  const httpUrl =
    opts.httpUrl || SWARM_HTTP_URL || 'https://onyx-storage.mainframe.com'
  const wsUrl = opts.wsUrl || SWARM_WS_URL || 'ws://localhost:8546'
  const certsDir = opts.certsDir || 'certs'
  let port = opts.port
  if (port == null) {
    port = ONYX_PORT == null ? 5000 : parseInt(ONYX_PORT, 10)
  }

  // Setup smart contracts (defaults to Mainnet)
  const ethNetwork = opts.testNet ? 'TESTNET' : 'MAINNET'
  const web3Url = opts.web3Url || WEB3_URLS[ethNetwork]
  const stakeEns = opts.stakeEnsAddress || ENS_STAKE_NAME[ethNetwork]
  const ensAddress = opts.ensAddress || ENS_ADDRESSES[ethNetwork]
  const contracts = createContracts(web3Url, stakeEns, ensAddress)
  // Setup DB using provided store (optional)
  const db = new DB(contracts, opts.store, `onyx-server-${port}`)
  // Connect to local Swarm node, this also makes the node's address and public key available in the db module
  const pss = await setupPss(db, wsUrl)

  // Derive wallet address from public key (stored as profile ID during setup)
  const profile = db.getProfile()
  if (profile == null) {
    throw new Error('Invalid setup')
  }
  const addr = pubKeyToAddress(profile.id)

  // Check if address has stake, throw otherwise
  const addrHasStake = await db.contracts.hasStake(addr)
  if (!addrHasStake) {
    const err: Object = new Error(`Missing stake for address ${addr}`)
    err.address = addr
    throw err
  }

  // Start listening to the "contact request" topic and handle these requests
  await setupContactTopic(pss, db)
  // Set subscriptions for stored convos
  await subscribeToStoredConvos(pss, db)
  // Start the BZZ and GraphQL server
  const server = await createServer(
    pss,
    db,
    httpUrl,
    port,
    !opts.unsecure,
    certsDir,
  )
  return server
}

export default start
